ADAL/src/request/ADAuthenticationRequest+AcquireToken.m

```diff
@@ -577,4 +547,143 @@ - (void)tryRefreshToken:(ADAuthenticationCallback)completionBlock
+        [result setCloudAuthority:_cloudAuthority];
+        completionHandler(result);
+        return YES;
+    }
+    
+    [ADBrokerHelper promptBrokerInstall:[NSURL URLWithString:authResponse.appInstallLink]
+                          brokerRequest:brokerRequestURL
+                      completionHandler:completionHandler];
+    return YES;
+}
+#endif
+
+- (BOOL)processOpenBrowserResponse:(MSIDWebviewResponse *)response
+                    telemetryEvent:(ADTelemetryAPIEvent *)event
+                 completionHandler:(ADAuthenticationCallback)completionHandler
+{
+    if (![response isKindOfClass:MSIDWebOpenBrowserResponse.class])
+    {
+        return NO;
+    }
+    
+    NSURL *browserURL = ((MSIDWebOpenBrowserResponse *)response).browserURL;
+    
+    
+#if TARGET_OS_IPHONE
+    if (![MSIDAppExtensionUtil isExecutingInAppExtension])
+    {
+        MSID_LOG_INFO(nil, @"Opening a browser");
+        MSID_LOG_INFO_PII(nil, @"Opening a browser - %@", browserURL);
+        dispatch_async( dispatch_get_main_queue(), ^{
+            [MSIDAppExtensionUtil sharedApplicationOpenURL:browserURL];
+        });
+        
+        [MSIDAppExtensionUtil sharedApplicationOpenURL:browserURL];
+    }
+    else
+    {
+        ADAuthenticationError *error = [ADAuthenticationError errorWithDomain:ADAuthenticationErrorDomain
+                                                                         code:AD_ERROR_UI_NOT_SUPPORTED_IN_APP_EXTENSION
+                                                            protocolErrorCode:nil
+                                                                 errorDetails:ADInteractionNotSupportedInExtension
+                                                                correlationId:_requestParams.correlationId];
+        
+        ADAuthenticationResult *result = [ADAuthenticationResult resultFromError:error correlationId:_requestParams.correlationId];
+        
+        [event setAPIStatus: MSID_TELEMETRY_VALUE_FAILED];
+        [[MSIDTelemetry sharedInstance] stopEvent:_requestParams.telemetryRequestId event:event];
+        
+        completionHandler(result);
+        return YES;
+    }
+#else
+    [[NSWorkspace sharedWorkspace] openURL:browserURL];
+#endif
+    ADAuthenticationResult *result = [ADAuthenticationResult resultFromCancellation:_requestParams.correlationId];
+    
+    [event setAPIStatus: MSID_TELEMETRY_VALUE_CANCELLED];
+    [[MSIDTelemetry sharedInstance] stopEvent:_requestParams.telemetryRequestId event:event];
+    
+    completionHandler(result);
+    return YES;
+}
+
+- (BOOL)processOAuthResponse:(MSIDWebviewResponse *)response
+               telemetryEvent:(ADTelemetryAPIEvent *)event
+            completionHandler:(ADAuthenticationCallback)completionHandler
+{
+    if (![response isKindOfClass:MSIDWebOAuth2Response.class])
+    {
+        return NO;
+    }
+    
+    MSIDWebOAuth2Response *oauthResponse = (MSIDWebOAuth2Response *)response;
+    
+    if (oauthResponse.authorizationCode)
```

Comment:
nit: should we check if it's empty too? (msidIsNilOrEmpty)

Reply:
I think it would be better handled at common, 
https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/223